File: server/internal/proto/events.go
// Package proto：访客事件类型（discriminated union）。
//
// 切片 1b 仅含：mouse_move / click / scroll / form_submit。
// 切片 1c 起加 rrweb 全量事件（IncrementalSnapshot / FullSnapshot 等）。
package proto

// EventType 是 visitor 事件类型。
type EventType string

const (
	EvMouseMove   EventType = "mouse_move"
	EvClick       EventType = "click"
	EvScroll      EventType = "scroll"
	EvFormSubmit  EventType = "form_submit"
	EvRRWeb       EventType = "rrweb" // 切片 1c 新增
)

// EventPayload 是 MsgEvent 的 payload（含事件类型与具体数据）。
// 1b 实现四类事件，1c 加 RRWeb 字段。
// 1c SDK 完全替换 4 类 collector 为 rrweb，1b 字段保留以便兼容历史 blob 回放。
type EventPayload struct {
	Type       EventType       `msgpack:"type"`
	TS         int64           `msgpack:"ts"` // 事件原始时间（毫秒）
	MouseMove  *MouseMoveData  `msgpack:"mouse_move,omitempty"`
	Click      *ClickData      `msgpack:"click,omitempty"`
	Scroll     *ScrollData     `msgpack:"scroll,omitempty"`
	FormSubmit *FormSubmitData `msgpack:"form_submit,omitempty"`
	RRWeb      *RRWebEvent     `msgpack:"rrweb,omitempty"` // 1c
}

// RRWebEvent 是单个 rrweb 事件（与 rrweb v2 SnapshotEvent 对应）。
// Data 字段类型复杂（rrweb-snapshot 内部类型），1c 用 raw bytes 透传不类型化。
type RRWebEvent struct {
	Type      int             `msgpack:"type"`      // rrweb 事件类型枚举（FullSnapshot=2, IncrementalSnapshot=3, Meta=4）
	Timestamp int64           `msgpack:"timestamp"` // rrweb 原始时间戳（毫秒）
	Data      map[string]any  `msgpack:"data"`      // rrweb data 字段（含 node tree、source、positions 等）
}

// MouseMoveData 是 mouse_move 事件的 data。
type MouseMoveData struct {
	X int `msgpack:"x"`
	Y int `msgpack:"y"`
}

// ClickData 是 click 事件的 data。
type ClickData struct {
	X              int    `msgpack:"x"`
	Y              int    `msgpack:"y"`
	Button         int    `msgpack:"button"` // 0=左, 1=中, 2=右
	TargetSelector string `msgpack:"target_selector,omitempty"`
}

// ScrollData 是 scroll 事件的 data。
type ScrollData struct {
	X int `msgpack:"x"`
	Y int `msgpack:"y"`
}

// FormSubmitData 是 form_submit 事件的 data。
type FormSubmitData struct {
	FormID string            `msgpack:"form_id,omitempty"`
	Fields map[string]string `msgpack:"fields"`
}

File: packages/proto/src/events.ts
// 访客事件类型（discriminated union，与 admin/server 同源）
// 1b 含：mouse_move / click / scroll / form_submit
// 1c 加：rrweb（覆盖前面 4 类）

export type EventType = 'mouse_move' | 'click' | 'scroll' | 'form_submit' | 'rrweb';

export interface MouseMoveData {
  x: number;
  y: number;
}

export interface ClickData {
  x: number;
  y: number;
  button: number;
  target_selector?: string;
}

export interface ScrollData {
  x: number;
  y: number;
}

export interface FormSubmitData {
  form_id?: string;
  fields: Record<string, string>;
}

/**
 * RRWebEvent 与 rrweb v2 的 SnapshotEvent 对应。
 * data 字段类型复杂，保留为 unknown 透传。
 */
export interface RRWebEvent {
  type: number; // rrweb 事件类型枚举：FullSnapshot=2, IncrementalSnapshot=3, Meta=4
  timestamp: number;
  data: unknown;
}

// 1b 用组合字段（与 Go 端 EventPayload 对应）
// 1c 加 rrweb 字段
export interface EventPayload {
  type: EventType;
  ts: number;
  mouse_move?: MouseMoveData;
  click?: ClickData;
  scroll?: ScrollData;
  form_submit?: FormSubmitData;
  rrweb?: RRWebEvent;
}

// 工厂函数：避免每次构造完整对象
export const mouseMove = (x: number, y: number, ts: number): EventPayload => ({
  type: 'mouse_move',
  ts,
  mouse_move: { x, y },
});

export const click = (
  x: number,
  y: number,
  button: number,
  targetSelector: string | undefined,
  ts: number,
): EventPayload => ({
  type: 'click',
  ts,
  click: { x, y, button, target_selector: targetSelector },
});

export const scroll = (x: number, y: number, ts: number): EventPayload => ({
  type: 'scroll',
  ts,
  scroll: { x, y },
});

export const formSubmit = (
  formId: string | undefined,
  fields: Record<string, string>,
  ts: number,
): EventPayload => ({
  type: 'form_submit',
  ts,
  form_submit: { form_id: formId, fields },
});

export const rrwebEvent = (rrweb: RRWebEvent): EventPayload => ({
  type: 'rrweb',
  ts: rrweb.timestamp,
  rrweb,
});

/** 判断是否为 rrweb FullSnapshot 事件（用于服务端缓存识别） */
export const isFullSnapshot = (e: EventPayload): boolean =>
  e.type === 'rrweb' && e.rrweb?.type === 2;
